### PR TITLE
add new rule for detecting change of additionalProperties from True to False

### LINKED ISSUE #4: Implement detection rule for changing type of additional properties
# Change value of additionalProperties
## Rationale
If the object is defined with additionalProperties set to False then the object will not allow presence of properties not defined on the properties section of the object definition.
Changing the value of additionalProperties could lead to incompatible changes when the new specs define a less stringent requirement.
## Mitigation
TODO

The main goal is to identify changes like
* {...} -> {"additionalProperties": False, ...}
* {...} -> {"additionalProperties": {xxx}, ...}
* {"additionalProperties": {...}, ...} -> {"additionalProperties": {something different}, ...}
* etc

### PR DESCRIPTION
**Pull request by yyang08**
This PR fixes #4 by implementing a new rule. Added tests and documentation.

### CODE REVIEW

**Comment by yyang08** [comment]
It's currently failing Travis CI builds for python2.7. Do you know how I could fix this? @macisamuele

**Comment by macisamuele** [comment]
@yyang08 I'll be reviewing this PR either today or tomorrow.
Sorry for the long turnaround

About the test failures, I revamped #30 that fixed tests on all the environments ... so I would not be very concerned for now

File: docs/source/rules/REQ-E004.rst [hunk 1]
```diff
@@ -0,0 +1,39 @@
+[REQ-E004] - Changing additionalProperties to False for a request parameter
+=====================================================
```
Comment (macisamuele) [inline docs/source/rules/REQ-E004.rst:2]: nit. Extend the `=` to underling all the caption characters ;)

```suggestion
===========================================================================
```

File: docs/source/rules/REQ-E004.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+Rationale
+---------
+If the object is defined with additionalProperties set to False then the object will not allow presence of properties not defined on the properties section of the object definition. Changing additionalProperties from True to False makes objects sent from a client, that contain additional properties and were permitted by the "old" Swagger specs, to the server be considered invalid by the backend.
+
+Mitigation
+----------
+A possible mitigation could be to not set additionalProperties to False in the object schema.
```
Comment (macisamuele) [inline docs/source/rules/REQ-E004.rst:10]: I would rather be more explicit here mentioning the fact that there there is no mitigation for this, but rather a recommended approach going forward.

You can find an example of it on https://swagger-spec-compatibility.readthedocs.io/en/latest/rules/MIS-E002.html#mitigation

File: docs/source/rules/REQ-E004.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+Example
+-------
+Old Swagger Specs
+~~~~~~~~~~~~~~~~~
+
+.. literalinclude:: examples/REQ-E004/old.yaml
+   :name: Old Swagger Spec
+   :language: yaml
+   :emphasize-lines: 17-18
```
Comment (macisamuele) [inline docs/source/rules/REQ-E004.rst:20]: This is emphasising
```yaml
            property_2:
              type: string
```
I think that is a copy-paste leftover.
Please remove

```suggestion
```

File: docs/source/rules/REQ-E004.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+   :linenos:
+
+New Swagger Specs
+~~~~~~~~~~~~~~~~~
+
+.. literalinclude:: examples/REQ-E004/new.yaml
+   :name: New Swagger Spec
+   :language: yaml
+   :linenos:
```
Comment (macisamuele) [inline docs/source/rules/REQ-E004.rst:29]: Here is missing the emphasis on the spec change

```suggestion
   :emphasize-lines: 13
   :linenos:
```

**Review by macisamuele — CHANGES_REQUESTED** [review]
@yyang08  thanks a lot for your contribution.

I've left few marginal comments on the documentation side, addressed them this PR is good to be merged.

About the test failures I would not be too concerned as we're trying to revamp the testing environment and the failure is not caused by this PR.

I've seen already in Yelp org PRs to address py2 compatibility
* https://github.com/Yelp/swagger_spec_validator/pull/143
* https://github.com/Yelp/bravado/pull/466

File: docs/source/rules/REQ-E004.rst (hunk 2 above)
Comment (yyang08) [reply docs/source/rules/REQ-E004.rst]: That makes sense - I'll change the phrasing to be more explicit.

**Comment by yyang08** [comment]
@macisamuele Thanks for the review. I've updated the doc per your suggestion. Glad to contribute - it's a great tool and my company is adopting it for service API development.

**Comment by macisamuele** [comment]
Merging this as tests are green on my local machine.
Fixing the CI is addressed by #30 (but we're waiting for some final agreements).

**Pull request merged**